### PR TITLE
 resources: aws: ec2: Refactor longpollWatch

### PR DESCRIPTION
**Pull request by jonathangold**
Complete rewrite of Watch() for correctness and maintanability.

### CODE REVIEW

File: resources/aws_ec2.go [hunk 1]
```diff
@@ -503,71 +280,67 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+		},
+	}
+	// call on the appropriate waiter based on current state.
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
```
Comment (jonathangold) [inline resources/aws_ec2.go]: There's a race where this returns some milliseconds before the instance is available (according to my describeInstanceByName function,) so we get 4 or 5 extra "terminated" events. I know sleeping for a second isn't a great fix, but it _would_ work in this case. Better ideas?

**Comment by jonathangold** [comment]
https://github.com/purpleidea/mgmt/pull/284/commits/61e5392b01a3c45d9da102f8c0abc9071fb22cc8 is just there so you can look at the additions on their own in https://github.com/purpleidea/mgmt/pull/284/commits/79cabc7f3cbcfa2bc4aced762289f98c2125c452 since this is a complete rewrite. https://github.com/purpleidea/mgmt/pull/284/commits/61e5392b01a3c45d9da102f8c0abc9071fb22cc8 contains all the deletions (+ 3 lines to make it "correct"). Now I'm going to bed :P

**Comment by jonathangold** [comment]
@purpleidea Got rid of the deletions.

File: resources/aws_ec2.go [hunk 2]
```diff
@@ -555,6 +491,73 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+		if !ok {
+			return "", errwrap.Wrapf(err, "error casting awserr")
+		}
+		// ignore these errors
+		if aerr.Code() != request.CanceledErrorCode && aerr.Code() != request.WaiterResourceNotReadyErrorCode {
+			return "", errwrap.Wrapf(err, "internal waiter error")
+		}
+		// if the waiter times out, start again recursively.
+		if aerr.Message() == AwsErrExceededWaitAttempts {
+			return watchWaiter(ctx, instance, c)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: While this might logically make sense, we shouldn't push code that depends on an infinitely deep recursion call stack, since *eventually* we would run out of memory. Not that it's necessarily likely to happen (I didn't measure) but it's logically inappropriate. So we need a different mechanism please.

File: resources/aws_ec2.go [hunk 3]
```diff
@@ -379,174 +388,101 @@ func (obj *AwsEc2Res) Watch() error {
 
 // longpollWatch uses the ec2 api's built in methods to watch ec2 resource state.
 func (obj *AwsEc2Res) longpollWatch() error {
-	send := false
+	var send bool
 	var exit *error
+
+	// we tell the engine we're running right away. Because watchWaiter
+	// checks the instance state directly so we don't miss any events.
 	if err := obj.Running(); err != nil {
```
Comment (jonathangold) [inline resources/aws_ec2.go]: I'd like to discuss this further, because I imagine when it comes time for send/recv this might no longer be race-free (if it is now.) Also, typo in the comment /s/,

File: resources/aws_ec2.go [hunk 4]
```diff
@@ -379,174 +388,101 @@ func (obj *AwsEc2Res) Watch() error {
...
 		return err
 	}
-	defer obj.wg.Wait()
-	defer close(obj.closeChan)
+
+	// channels and waitgroup for goroutines.
```
Comment (jonathangold) [inline resources/aws_ec2.go]: I forgot to change these back to the struct channels and wg. I can change it back after you review.

File: resources/aws_ec2.go (hunk 2 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: OK. Good thing we didn't go with HASKELL :P 
Do you want me to fix this before you review?

**Comment by jonathangold** [comment]
@purpleidea fixed the recursion.

File: resources/aws_ec2.go [hunk 5]
```diff
@@ -555,6 +486,73 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if !ok {
+			return "", errwrap.Wrapf(err, "error casting awserr")
+		}
+		// ignore these errors
+		if aerr.Code() != request.CanceledErrorCode && aerr.Code() != request.WaiterResourceNotReadyErrorCode {
+			return "", errwrap.Wrapf(err, "internal waiter error")
+		}
+		// if the waiter times out, start again recursively.
```
Comment (jonathangold) [inline resources/aws_ec2.go]: Forgot to change the comment. Will do after you review.

File: resources/aws_ec2.go (hunk 1 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: This waiter doesn't work the way I thought it did. Troubleshooting it now.

Comment (jonathangold) [reply resources/aws_ec2.go]: Figured it out, I was waiting for any instance with the right name to exist, including terminated ones. ಠ_ಠ
I think I made a change that stopped it waiting for terminated ones in the original version, which might be why waituntilinstancestopped wasn't behaving how I thought.

File: resources/aws_ec2.go [hunk 6]
```diff
@@ -374,19 +383,28 @@ func (obj *AwsEc2Res) Watch() error {
...
+	var send bool
 	var exit *error
+
+	// we tell the engine we're running right away. Because watchWaiter
+	// checks the instance state directly so we don't miss any events.
 	if err := obj.Running(); err != nil {
 		return err
 	}
+
+	// canellable context used for exiting cleanly.
```
Comment (purpleidea) [inline resources/aws_ec2.go]: typo. one liners are either full sentences with capitalization and period or have neither same above and in a bunch of other comments.

File: resources/aws_ec2.go [hunk 7]
```diff
@@ -374,19 +383,28 @@ func (obj *AwsEc2Res) Watch() error {
...
-	return obj.longpollWatch()
+	return obj.waiterWatch()
 }
 
-// longpollWatch uses the ec2 api's built in methods to watch ec2 resource state.
-func (obj *AwsEc2Res) longpollWatch() error {
-	send := false
+// waiterWatch uses the ec2 api's built in methods to watch ec2 resource state.
+func (obj *AwsEc2Res) waiterWatch() error {
+	var send bool
```
Comment (purpleidea) [inline resources/aws_ec2.go]: why are we renaming this and also we are we making these sort of unrelated changes in the same commit?

`send := false`
is functionally the same as
`var send bool`

File: resources/aws_ec2.go [hunk 8]
```diff
@@ -555,6 +486,82 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+	}
+	// call on the appropriate waiter based on current state.
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		waitInput.Filters = append(waitInput.Filters, existWaiterFilter)
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: default state should be an error or a panic or ???

File: resources/aws_ec2.go [hunk 9]
```diff
@@ -374,19 +383,28 @@ func (obj *AwsEc2Res) Watch() error {
...
 
-// longpollWatch uses the ec2 api's built in methods to watch ec2 resource state.
-func (obj *AwsEc2Res) longpollWatch() error {
-	send := false
+// waiterWatch uses the ec2 api's built in methods to watch ec2 resource state.
+func (obj *AwsEc2Res) waiterWatch() error {
+	var send bool
 	var exit *error
+
+	// we tell the engine we're running right away. Because watchWaiter
```
Comment (purpleidea) [inline resources/aws_ec2.go]: So the fact that we previously did this right away was wrong, and now we've got a comment saying it's right. It was wrong before, but it was a bug. Now the comment incorrectly re-enforces the bug. :/

obj.Running _tells_ the engine that we're now watching! This is a lie!! It's limited by their long poll API not exposing when they're ready, but still.

File: resources/aws_ec2.go [hunk 10]
```diff
@@ -555,6 +486,82 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+		if aws.StringValue(tag.Key) == nameKey {
+			name = aws.StringValue(tag.Value)
+		}
+	}
+	// error if we didn't find one.
+	if name == "" {
+		return "", fmt.Errorf("name not found")
+	}
+
+	// build the input for the waiters.
```
Comment (purpleidea) [inline resources/aws_ec2.go]: why are the following 40-50 lines correct??

File: resources/aws_ec2.go (hunk 7 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: I can leave it as longpollWatch, but IMO, longpolling and using the waiters aren't exactly the same, and I think it's causing some confusion. I can also change it back to `send := false`, (I just like `var send bool` better style wise and thought it was okay since this is a refactor)

File: resources/aws_ec2.go (hunk 6 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: I guess I picked this up as a bad habit somewhere along the way. Will correct all of them.

File: resources/aws_ec2.go (hunk 10 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: I will add some comments explaining the logic and the reasoning behind it. Sorry if it's unclear.

File: resources/aws_ec2.go (hunk 8 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: oops. I'll add an error.

File: resources/aws_ec2.go (hunk 9 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: I don't think I understand the potential consequences of this race. What exactly does this _mean_ to the engine and what could go wrong if we tell the engine we're watching a second or two before we poll the state and start the waiters?

Comment (jonathangold) [reply resources/aws_ec2.go]: IOW, is this _just_ about potentially missing events, or is there more to this race on the engine side?

**Comment by jonathangold** [comment]
@purpleidea I made the changes. I am confident in this code. So if you ack you can merge (and squash with the last commit if you don't mind) I still want to discuss the obj.Running() race so I have a better handle on how it affects this resource and future ones, but I updated the comment to reflect the fact that it isn't correct.

File: resources/aws_ec2.go [hunk 11]
```diff
@@ -555,6 +489,93 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		waitInput.Filters = append(waitInput.Filters, existWaiterFilter)
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
+	default:
+		return "", errwrap.Wrapf(err, "unrecognized instance state")
```
Comment (purpleidea) [inline resources/aws_ec2.go]: what is the err var set to here?

File: resources/aws_ec2.go [hunk 12]
```diff
@@ -381,12 +390,24 @@ func (obj *AwsEc2Res) Watch() error {
...
+	// We tell the engine that we're running right away. This is not correct,
+	// but the api doesn't have a way to signal when the waiters are ready.
+	// Because watchWaiter checks the state directly, and the waiters will
+	// return as expected even if they miss the state change, we shouldn't miss
+	// any events.
 	if err := obj.Running(); err != nil {
 		return err
 	}
+
+	// canellable context used for exiting cleanly
```
Comment (purpleidea) [inline resources/aws_ec2.go]: typo, cannelloni

File: resources/aws_ec2.go [hunk 13]
```diff
@@ -381,12 +390,24 @@ func (obj *AwsEc2Res) Watch() error {
 func (obj *AwsEc2Res) longpollWatch() error {
 	send := false
 	var exit *error
+
+	// We tell the engine that we're running right away. This is not correct,
+	// but the api doesn't have a way to signal when the waiters are ready.
+	// Because watchWaiter checks the state directly, and the waiters will
```
Comment (purpleidea) [inline resources/aws_ec2.go]: i think these last 3 lines are incorrect and we should drop them

File: resources/aws_ec2.go [hunk 14]
```diff
@@ -555,6 +489,93 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		waitInput.Filters = append(waitInput.Filters, existWaiterFilter)
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
+	default:
+		return "", fmt.Errorf("unrecognized instance state")
```
Comment (purpleidea) [inline resources/aws_ec2.go]: it's customary to sometimes say which one, with a %s at the end, eg:

fmt.Errorf("unrecognized instance state: %s", blah)

or %+v if it's a weird value that's not a string

**Comment by jonathangold** [comment]
@purpleidea Fixed. :D

**Comment by purpleidea** [comment]
This is now YOLO merged :)

(I didn't review it 100% but I'll leave it to you!)

Thanks!
